File: lib/executors/remote.ts
/**
 * Remote Executor
 *
 * Executes agent commands on a remote VM via SSH + Docker.
 * Ideal for always-on execution accessible from anywhere (including mobile).
 *
 * Features:
 * - SSH tunnel to remote machine
 * - Docker containers on remote host
 * - tmux for interactivity (attach from anywhere)
 * - Project files synced via rsync or mounted
 *
 * Prerequisites:
 * - SSH access to remote VM (key-based auth recommended)
 * - Docker installed on remote VM
 * - orchestra-agent image available on remote
 */

import { spawn } from 'child_process';
import type { ExecutionResult, DockerConfig, RemoteConfig, ExecuteRequest } from '../types';
import { buildCommand } from './local';
import { escapeShellArg } from './shell';

// Default Docker image for agents
const DEFAULT_IMAGE = 'orchestra-agent:full';

// SSH connection timeout
const SSH_TIMEOUT_SECONDS = 30;

/**
 * Execute an agent command on a remote VM
 */
export async function executeRemote(request: ExecuteRequest): Promise<ExecutionResult> {
  const { executor, prompt, options, executionConfig, projectPath, nodeId } = request;
  const remoteConfig = executionConfig?.remote;
  const dockerConfig = executionConfig?.docker || {};
  const interactiveConfig = executionConfig?.interactive || {};

  if (!remoteConfig?.host) {
    return {
      status: 'error',
      error: 'Remote host is required for remote execution',
    };
  }

  // Generate session name
  const sessionName = interactiveConfig.sessionName || `agent-${nodeId || Date.now()}`;
  const containerName = `orchestra-${sessionName}`;

  // Build the agent command
  const agentArgs = buildCommand(executor as 'claude' | 'codex' | 'gemini', prompt, options);
  const agentCommand = agentArgs.map(escapeShellArg).join(' ');

  try {
    // Sync project files to remote if projectPath is provided
    if (projectPath) {
      await syncProjectToRemote(projectPath, remoteConfig, sessionName);
    }

    // Build and execute the remote Docker command
    const dockerCommand = buildRemoteDockerCommand(
      dockerConfig,
      containerName,
      sessionName,
      agentCommand
    );

    await executeSSHCommand(remoteConfig, dockerCommand);

    // Build attach command
    const attachCommand = buildAttachCommand(remoteConfig, containerName);

    return {
      status: 'running',
      sessionId: containerName,
      attachCommand,
    };
  } catch (error) {
    return {
      status: 'error',
      error: error instanceof Error ? error.message : String(error),
    };
  }
}

/**
 * Check remote session status
 */
export async function getRemoteSessionStatus(
  remoteConfig: RemoteConfig,
  sessionId: string
): Promise<'running' | 'stopped' | 'unknown'> {
  try {
    const output = await executeSSHCommand(
      remoteConfig,
      `docker inspect -f '{{.State.Running}}' ${sessionId} 2>/dev/null || echo 'not_found'`
    );

    if (output.includes('true')) return 'running';
    if (output.includes('false')) return 'stopped';
    return 'unknown';
  } catch {
    return 'unknown';
  }
}

/**
 * Get output from remote session
 */
export async function getRemoteSessionOutput(
  remoteConfig: RemoteConfig,
  sessionId: string
): Promise<string> {
  return executeSSHCommand(
    remoteConfig,
    `docker exec ${sessionId} tmux capture-pane -t agent -p -S -10000 2>/dev/null || docker logs ${sessionId} 2>&1`
  );
}

/**
 * Stop a remote session
 */
export async function stopRemoteSession(
  remoteConfig: RemoteConfig,
  sessionId: string
): Promise<void> {
  await executeSSHCommand(remoteConfig, `docker stop ${sessionId}`);
}

/**
 * List all Orchestra sessions on remote
 */
export async function listRemoteSessions(
  remoteConfig: RemoteConfig
): Promise<Array<{ id: string; status: string; created: string }>> {
  const output = await executeSSHCommand(
    remoteConfig,
    `docker ps -a --filter "name=orchestra-" --format "{{.Names}}|{{.Status}}|{{.CreatedAt}}"`
  );

  return output
    .split('\n')
    .filter(Boolean)
    .map((line) => {
      const [id, status, created] = line.split('|');
      return { id, status, created };
    });
}

// ========== INTERNAL HELPERS ==========

function buildSSHArgs(config: RemoteConfig): string[] {
  const args: string[] = [];

  // SSH key
  if (config.keyPath) {
    args.push('-i', config.keyPath);
  }

  // Port
  if (config.port) {
    args.push('-p', String(config.port));
  }

  // Connection timeout
  args.push('-o', `ConnectTimeout=${SSH_TIMEOUT_SECONDS}`);

  // Disable strict host key checking for convenience (can be made configurable)
  args.push('-o', 'StrictHostKeyChecking=accept-new');

  // User@Host
  const user = config.user || 'root';
  args.push(`${user}@${config.host}`);

  return args;
}

function executeSSHCommand(config: RemoteConfig, command: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const sshArgs = [...buildSSHArgs(config), command];

    const proc = spawn('ssh', sshArgs, {
      shell: false,
      stdio: ['ignore', 'pipe', 'pipe'],
    });

    let stdout = '';
    let stderr = '';

    proc.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    proc.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    proc.on('close', (code) => {
      if (code === 0) {
        resolve(stdout.trim());
      } else {
        reject(new Error(stderr || `SSH command failed with code ${code}`));
      }
    });

    proc.on('error', reject);
  });
}

async function syncProjectToRemote(
  localPath: string,
  config: RemoteConfig,
  sessionName: string
): Promise<void> {
  const user = config.user || 'root';
  const remotePath = `/tmp/orchestra-projects/${sessionName}`;

  // Create remote directory
  await executeSSHCommand(config, `mkdir -p ${remotePath}`);

  // Rsync files
  return new Promise((resolve, reject) => {
    const rsyncArgs = [
      '-avz',
      '--delete',
      '-e',
      `ssh ${config.keyPath ? `-i ${config.keyPath}` : ''} ${config.port ? `-p ${config.port}` : ''}`,
      `${localPath}/`,
      `${user}@${config.host}:${remotePath}/`,
    ];

    const proc = spawn('rsync', rsyncArgs, {
      shell: false,
      stdio: ['ignore', 'pipe', 'pipe'],
    });

    let stderr = '';

    proc.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    proc.on('close', (code) => {
      if (code === 0) {
        resolve();
      } else {
        reject(new Error(stderr || `Rsync failed with code ${code}`));
      }
    });

    proc.on('error', reject);
  });
}

function buildRemoteDockerCommand(
  config: DockerConfig,
  containerName: string,
  sessionName: string,
  agentCommand: string
): string {
  const parts: string[] = ['docker', 'run', '-d'];

  // Container name
  parts.push('--name', containerName);

  // Resource limits
  if (config.resources?.memory) {
    parts.push('--memory', config.resources.memory);
  }
  if (config.resources?.cpus) {
    parts.push('--cpus', config.resources.cpus);
  }

  // Network
  if (config.network) {
    parts.push('--network', config.network);
  }

  // Mount project directory (using the synced path)
  parts.push('-v', `/tmp/orchestra-projects/${sessionName}:/workspace`);
  parts.push('-w', '/workspace');

  // Environment variables - these should be set on the remote host
  // or passed through SSH environment forwarding
  parts.push('-e', 'CLAUDE_CODE_OAUTH_TOKEN');
  parts.push('-e', 'ANTHROPIC_API_KEY');
  parts.push('-e', 'OPENAI_API_KEY');
  parts.push('-e', 'GOOGLE_API_KEY');

  // Image
  parts.push(config.image || DEFAULT_IMAGE);

  // Start tmux with the agent command
  parts.push('tmux', 'new-session', '-d', '-s', 'agent', agentCommand);

  return parts.join(' ');
}

function buildAttachCommand(config: RemoteConfig, containerName: string): string {
  const user = config.user || 'root';
  const keyArg = config.keyPath ? `-i ${config.keyPath}` : '';
  const portArg = config.port ? `-p ${config.port}` : '';

  return `ssh -t ${keyArg} ${portArg} ${user}@${config.host} "docker exec -it ${containerName} tmux attach -t agent"`;
}

/**
 * Check SSH connectivity to remote host
 */
export async function checkRemoteConnectivity(config: RemoteConfig): Promise<boolean> {
  try {
    await executeSSHCommand(config, 'echo "ok"');
    return true;
  } catch {
    return false;
  }
}

/**
 * Check if Docker is available on remote host
 */
export async function checkRemoteDocker(config: RemoteConfig): Promise<boolean> {
  try {
    await executeSSHCommand(config, 'docker version');
    return true;
  } catch {
    return false;
  }
}

/**
 * Check if the agent image is available on remote host
 */
export async function checkRemoteImage(
  config: RemoteConfig,
  image: string = DEFAULT_IMAGE
): Promise<boolean> {
  try {
    await executeSSHCommand(config, `docker image inspect ${image}`);
    return true;
  } catch {
    return false;
  }
}

File: src-tauri/src/executors/remote.rs
//! Remote executor - runs agents on remote VMs via SSH

use super::{ExecuteRequest, ExecutionResult, ExecutorError, ExecutorResult};
use std::process::Stdio;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use tokio::time::{timeout, Duration};

/// Execution timeout (15 minutes for remote)
const EXECUTION_TIMEOUT: Duration = Duration::from_secs(15 * 60);

/// Execute an agent on a remote VM via SSH + Docker
pub async fn execute_remote<F>(
    request: &ExecuteRequest,
    on_output: F,
) -> ExecutorResult<ExecutionResult>
where
    F: Fn(String) + Send + 'static,
{
    let remote_config = request
        .execution_config
        .as_ref()
        .and_then(|c| c.remote.as_ref())
        .ok_or_else(|| ExecutorError::Remote("Remote config required".to_string()))?;

    let host = &remote_config.host;
    let user = remote_config.user.as_deref().unwrap_or("root");
    let port = remote_config.port.unwrap_or(22);

    // Build SSH connection arguments
    let mut ssh_args = vec![
        "-o".to_string(),
        "StrictHostKeyChecking=no".to_string(),
        "-o".to_string(),
        "BatchMode=yes".to_string(),
        "-p".to_string(),
        port.to_string(),
    ];

    if let Some(key_path) = &remote_config.key_path {
        ssh_args.push("-i".to_string());
        ssh_args.push(key_path.clone());
    }

    ssh_args.push(format!("{}@{}", user, host));

    // Build the remote Docker command
    let docker_config = request
        .execution_config
        .as_ref()
        .and_then(|c| c.docker.as_ref());

    let image = docker_config
        .and_then(|c| c.image.as_ref())
        .map(|s| s.as_str())
        .unwrap_or("orchestra-agent:full");

    let agent_command = build_agent_command(&request.executor, &request.prompt, &request.options);

    // Build Docker run command for remote
    let docker_command = format!(
        "docker run --rm {} sh -c {}",
        image,
        shell_escape(&agent_command)
    );

    ssh_args.push(docker_command);

    tracing::info!("Executing on remote: ssh {}@{}", user, host);

    // Spawn SSH process
    let mut child = Command::new("ssh")
        .args(&ssh_args)
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .map_err(|e| ExecutorError::Remote(format!("Failed to start SSH: {}", e)))?;

    // Stream output
    let stdout = child.stdout.take().unwrap();
    let stderr = child.stderr.take().unwrap();

    let mut stdout_reader = BufReader::new(stdout).lines();
    let mut stderr_reader = BufReader::new(stderr).lines();

    let mut output = String::new();

    let result = timeout(EXECUTION_TIMEOUT, async {
        loop {
            tokio::select! {
                line = stdout_reader.next_line() => {
                    match line {
                        Ok(Some(line)) => {
                            output.push_str(&line);
                            output.push('\n');
                            on_output(format!("{}\n", line));
                        }
                        Ok(None) => break,
                        Err(_) => break,
                    }
                }
                line = stderr_reader.next_line() => {
                    match line {
                        Ok(Some(line)) => {
                            // SSH stderr might contain connection info, still output it
                            output.push_str(&line);
                            output.push('\n');
                            on_output(format!("{}\n", line));
                        }
                        Ok(None) => {}
                        Err(_) => {}
                    }
                }
            }
        }

        child.wait().await
    })
    .await;

    match result {
        Ok(Ok(status)) => {
            if status.success() {
                Ok(ExecutionResult::Done { output })
            } else {
                Ok(ExecutionResult::Error {
                    message: format!("Remote execution failed with code {}", status.code().unwrap_or(-1)),
                })
            }
        }
        Ok(Err(e)) => Err(ExecutorError::Remote(e.to_string())),
        Err(_) => {
            let _ = child.kill().await;
            Err(ExecutorError::Timeout)
        }
    }
}

/// Build the agent command string
fn build_agent_command(
    executor: &str,
    prompt: &str,
    options: &Option<serde_json::Value>,
) -> String {
    let escaped_prompt = shell_escape(prompt);

    match executor {
        "claude" => {
            let mut cmd = format!(
                "claude -p {} --output-format text --no-session-persistence --permission-mode dontAsk --tools ''",
                escaped_prompt
            );

            if let Some(opts) = options {
                if let Some(model) = opts.get("model").and_then(|v| v.as_str()) {
                    cmd.push_str(&format!(" --model {}", model));
                }
            }

            cmd
        }

        "codex" => {
            let mut cmd = "codex exec --skip-git-repo-check".to_string();

            if let Some(opts) = options {
                let reasoning = opts
                    .get("reasoningEffort")
                    .or_else(|| opts.get("reasoningLevel"))
                    .and_then(|v| v.as_str());

                if let Some(level) = reasoning {
                    cmd.push_str(&format!(" -c reasoning.effort={}", level));
                }
            }

            cmd.push_str(&format!(" {}", escaped_prompt));
            cmd
        }

        "gemini" => {
            let model = options
                .as_ref()
                .and_then(|o| o.get("model"))
                .and_then(|v| v.as_str())
                .unwrap_or("gemini-3-pro-preview");

            format!("gemini {} -m {} -o text", escaped_prompt, model)
        }

        _ => escaped_prompt,
    }
}

/// Escape a string for shell use
fn shell_escape(s: &str) -> String {
    format!("'{}'", s.replace("'", "'\\''"))
}
